oeps/oep-0041-arch-async-server-event-messaging.rst

```diff
@@ -0,0 +1,486 @@
+  exceeded the required threshold.
+
+Open edX services already use `Django signals
+<https://docs.djangoproject.com/en/1.11/topics/signals/>`_ for subscribing to
+events within the same service, but we have no accepted recommendation for how
+to do something similar across service boundaries. As the number of services in
+Open edX has grown, the various ad hoc synchronization and communication methods
+we've implemented over the years have caused numerous operational and
+performance problems (particularly around catalog data). At the same time,
+recent discussions around Domain Driven Design and data replication have given
+us a better framing for the kind of relationship we want to have between various
+services. There has also been a strong desire for a properly documented set of
+integration APIs that would allow Open edX to better interoperate with other
+applications in the ecosystem.
+
+There is an opportunity to bring all these elements together and create a solid
+foundation for simple, reliable, well documented, near real-time cross-service
+message passing. But without some planning and design work, we may end up in an
+even more chaotic place where each pairing of services uses different
+conventions for this kind of communication.
+
+
+-------------
+Specification
+-------------
+
+Message Format
+==============
+
+The message format will use the `CloudEvents 1.0
+<https://github.com/cloudevents/spec/blob/master/spec.md>`_ specification, with
+additional usage conventions that all Open edX services should adhere to.
+CloudEvents was developed by the Cloud Native Computing Foundation, with
+participation from major companies like Google, Microsoft, IBM, Red Hat, Oracle,
+and Huawei. While the finalized 1.0 spec is only months old and the Python
+tooling is primitive (advertised to break at any time), CloudEvents is also a
+pretty minimal message envelope spec, with the basic required fields being
+things that we'd want to have anyway.
+
+Event messages will be UTF-8 encoded JSON documents of 64 KB or less. An
+example:
+
+.. code-block:: javascript
+
+  {
+    "id": "b3b3981e-7544-11ea-b663-acde48001122",
+    "source": "https://discovery.edx.org/events/"
+    "specversion": "1.0",
+    "type": "org.openedx.catalog.course.created.v1"
+    "time": "2020-02-23T09:43:18Z",
+    "minorversion": 0,
+    "data": {
+        // message specific payload here
+    }
+  }
+
+CloudEvents Fields
+==================
+
+All top level field names are specified by the CloudEvents spec, but most fields
+defined by that standard have some leeway in how they're used for any given
+system. Each field below has a link to CloudEvents requirements, as well as a
+description of the conventions Open edX will use for that field. All Open edX
+events should have all of the fields specified below (even if some are
+considered optional in the CloudEvents spec).
+
+`id <https://github.com/cloudevents/spec/blob/master/spec.md#id>`_
```

Comment:
The heading for each CloudEvents Field is inconsistent; some headers are hyperlinked and some aren't.